### PR TITLE
Search collection fixes

### PR DESCRIPTION
**Pull request by Atticus29**
# Description

This PR adds some improvements to the new search form. Specifically, it:

- Improves the spacing between the sections for trait-based searching section
-  Adds a trait search section when the configuration file specifies that one is needed
- Adds data-chips and relevant onchange listeners to the trait checkboxes and various select options, so that chips will appear upon selections thereof
- Adds logic that handles removal of the criteria chips for taxontype and materialsampletype select options
- Adds a button at the top of the accordion set that can collapse or expand all accordions
- Rearranges the order of the accordions to represent the most common use cases up top
- Fixes bug where map selector links were broken in the latitude and longitude accordion
- Adds some new translations where necessary and fixes a misformatted one.

# QA Recommendations - ping Mark when this is ready for QA, and he will set this branch up on one of the dev servers.

- Select and deselect multiple trait criteria from the accordion.
- Try removing criteria chips from the Criteria section and confirming that resulting searches result in correct behavior
- Change taxon type in the taxon accordion and material sample type from the sample properties accordion and confirm the correct search results.
- Try removing criteria chips from the Criteria section for material sample type and taxontype as well.
- The accordions were at one time a little buggy in terms of playing well with the "expand all sections"/"collapse all sections" button. Clicking individual accordions should not exempt them from the expand/collapse behavior. Confirm that there is no regression with this bug fix.

# Pull Request Checklist:

# Pre-Approval

- [x] There is a description section in the pull request that details what the proposed changes do. It can be very brief if need be, but it ought to exist.
- [ ] Hotfixes should be branched off of the `master` branch and **squash and merged** back into the `master` branch.
    - N/A
- [x] Features and backlog bugs should be merged into the `Development` branch, **NOT** `master`
- [x] All new text is preferably internationalized (i.e., no end-user-visible text is hard-coded on the PHP pages), and the [spreadsheet tracking internationalizations](https://docs.google.com/spreadsheets/d/133fps9w2pUCEjUA6IGCcQotk7dn9KvepMXJ2IWUZsE8/edit?usp=sharing) has been updated either with a new row or with checkmarks to existing rows.
- [x] There are no linter errors
- [x] New features have responsive design (i.e., look aesthetically pleasing both full screen and with small or mobile screens)
- [x] [Symbiota coding standards](https://docs.google.com/document/d/1-FwCZP5Zu4f-bPwsKeVVsZErytALOJyA2szjbfSUjmc/edit?usp=sharing) have been followed
- [ ] If any files have been reformatted (e.g., by an autoformatter), the reformat is its own, separate commit in the PR
    - N/A
- [x] Comment which GitHub issue(s), if any does this PR address
    - [https://github.com/orgs/BioKIC/projects/6/views/10?pane=issue&itemId=52430289](https://github.com/orgs/BioKIC/projects/6/views/10?pane=issue&itemId=52430289)
- [ ] If this PR makes any changes that would require additional configuration of any Symbiota portals outside of the files tracked in this repository, make sure that those changes are detailed in [this document](https://docs.google.com/document/d/1T7xbXEf2bjjm-PMrlXpUBa69aTMAIROPXVqJqa2ow_I/edit?usp=sharing).
    - N/A

# Post-Approval

- [ ] It is the code author's responsibility to merge their own pull request after it has been approved
- [ ] If this PR represents a merge into the `Development` branch, remember to use the **squash & merge** option
- [ ] If this PR represents a merge from the `Development` branch into the master branch, remember to use the **merge** option
- [ ] If this PR represents a hotfix into the `master` branch, a subsequent PR from `master` into `Development` should be made **merge** option (i.e., no squash).
- [ ] If the dev team has agreed that this PR represents the last PR going into the `Development` branch before a tagged release (i.e., before an imminent merge into the master branch), make sure to notify the team and [lock the `Development` branch](https://github.com/BioKIC/Symbiota/settings/branches) to prevent accidental merges while QA takes place. Follow the release protocol [here](https://github.com/BioKIC/Symbiota/blob/master/docs/release-protocol.md).
- [ ] Don't forget to delete your feature branch upon merge. Ignore this step as required.

Thanks for contributing and keeping it clean!

### CODE REVIEW

File: classes/OccurrenceAttributeSearch.php [hunk 1]
```diff
@@ -59,10 +59,10 @@ private function getTraitSearchHTML($traitID,$classStr=''){
 			$classArr = explode(' ',$classStr);
 			$divClass = array_pop($classArr);
 		}
-		$retStr = '<div class="'.$divClass.'" style="margin-left:'.($classStr?'10':'').'px;"><div style="clear:both">';
+		$retStr = '<div class="' . $divClass . '" style="margin-left:' . ($classStr?'10':'') . 'px; margin-top: 0.5rem;"><div style="clear:both">';
```
Comment (Atticus29) [inline classes/OccurrenceAttributeSearch.php:62]: Inline because in a class file. 😬

File: classes/OccurrenceAttributeSearch.php [hunk 2]
```diff
@@ -59,10 +59,10 @@ private function getTraitSearchHTML($traitID,$classStr=''){
...
 		if(isset($this->traitArr[$traitID]['states'])){
 			if($this->traitArr[$traitID]['type']=='TF'){
-				$retStr .= '<div style="float:left;margin-left: 15px">'.$this->traitArr[$traitID]['name'].':</div>';
+				$retStr .= '<div style="float:left;margin-left: 15px; margin-bottom: 0.5rem;">' . $this->traitArr[$traitID]['name'] . ':</div>';
```
Comment (Atticus29) [inline classes/OccurrenceAttributeSearch.php:65]: Inline because in a class file. 😬

File: classes/OccurrenceAttributeSearch.php [hunk 3]
```diff
@@ -75,18 +75,18 @@ private function getTraitSearchHTML($traitID,$classStr=''){
 					continue;
 				}
 				else{
-					$retStr .= '<div title="'.$sArr['description'].'" style="clear:both">';
-					$retStr .= '<input name="attr[]" id="traitstateid-'.$sid.'" class="'.$classStr.'" type="checkbox" value="'.$sid.'" onchange="traitChanged(this)" /> ';
-					$retStr .= $sArr['name'];
+					$retStr .= '<div title="' . $sArr['description'] . '" style="clear:both;">';
+					$retStr .= '<input data-chip="' . $this->traitArr[$traitID]['name'] . ': ' . ucfirst($sArr['name']) . '" name="attr[]" id="traitstateid-' . $sid . '" class="' . $classStr . '" type="checkbox" value="' . $sid . '" onchange="traitChanged(this)" /> ';
```
Comment (Atticus29) [inline classes/OccurrenceAttributeSearch.php:79]: Adds a data-chip attribute to this. Essential for the new search page and harmless in harvestparams.php, which will share this code.

File: collections/search/index.php [hunk 4]
```diff
@@ -46,6 +47,29 @@
 	<script src="<?php echo $CLIENT_ROOT ?>/js/jquery-3.7.1.min.js" type="text/javascript"></script>
 	<script>
 		const clientRoot = '<?php echo $CLIENT_ROOT; ?>';
+		const handleAccordionExpand = () => {
```
Comment (Atticus29) [inline collections/search/index.php:49]: Happy to put this in a js file and not clutter this PHP file if desired.

File: collections/search/js/searchform.js [hunk 5]
```diff
@@ -111,7 +113,7 @@ function openCoordAid(mapMode) {
  * @param {HTMLObjectElement} element Input for which chips are going to be created by default
  */
 function addChip(element) {
-  if (!element || !element.name) return;
+  if (!element || (!element.name && element?.tagName !== "OPTION")) return;
```
Comment (Atticus29) [inline collections/search/js/searchform.js:116]: Exempt option elements from this return.

File: content/lang/collections/search/index.en.php [hunk 6]
```diff
@@ -108,5 +106,9 @@
 $LANG['RADIUS'] = '_Radius';
 $LANG['ALERT_MSG_PREVIOUS_SEARCH_FORM'] = 'Looking for the previous search form? You can still use it here:';
 $LANG['ALL_COLLECTIONS'] = 'All Collections';
+$LANG['TRIAT_CRITERIA'] = 'Trait Criteria';
```
Comment (themerekat) [inline content/lang/collections/search/index.en.php]: Typo?

File: collections/search/index.php [hunk 7]
```diff
@@ -425,6 +386,141 @@
...
+					if(!empty($SEARCH_BY_TRAITS)) {
+						$attribSearch = new OccurrenceAttributeSearch();
+						$traitArr = $attribSearch->getTraitSearchArr($SEARCH_BY_TRAITS);
+						if($traitArr){
+				?>
+							<section>
+								<!-- Accordion selector -->
+								<input type="checkbox" id="trait" class="accordion-selector" />
+								<!-- Accordion header -->
+								<label for="trait" class="accordion-header"><?php echo $LANG['TRIAT_CRITERIA'] ?></label>
```
Comment (themerekat) [inline collections/search/index.php]: Typo (although it's consistent in your lang files too, haha!)

**Review by themerekat — CHANGES_REQUESTED** [review]
Technically I think this works as is, but I'd recommend changing TRIAT to TRAIT.

Ready to test in a dev environment!

File: content/lang/collections/search/index.en.php (hunk 6 above)
Comment (Atticus29) [reply content/lang/collections/search/index.en.php]: Fixed in a490b5535. Good catch!

File: collections/search/index.php (hunk 7 above)
Comment (Atticus29) [reply collections/search/index.php]: a490b5535

**Comment by Atticus29** [comment]
> Technically I think this works as is, but I'd recommend changing TRIAT to TRAIT.
> 
> Ready to test in a dev environment!

@themerekat . Dev environment set up on https://dev001.symbiota.org/collections/search/index.php. I still don't quite know how to set up a portal for material samples yet, but everything else should be ready for QA!

**Comment by themerekat** [comment]
When you conduct a search, then return to your Search Criteria using the breadcrumbs:
image

All of your search criteria go away. In harvestparams, they stuck around so that you could refine your search instead of having to re-create it from scratch. I think this is a feature that people will want to keep.

That being said, I'm super excited to see that clicking the "back" button finally takes you to the search page without the scary "resubmit data" warning!

**Comment by themerekat** [comment]
The underscores are showing up in the user interface now at the beginning of some of the search fields
image

**Comment by Atticus29** [comment]
> When you conduct a search, then return to your Search Criteria using the breadcrumbs: image
> 
> All of your search criteria go away. In harvestparams, they stuck around so that you could refine your search instead of having to re-create it from scratch. I think this is a feature that people will want to keep.
> 
> That being said, I'm super excited to see that clicking the "back" button finally takes you to the search page without the scary "resubmit data" warning!

Ok. I fixed this in 0fd7569d1. However, in the process of troubleshooting, I noticed some bugs that already existed in the harvestparams version as well. Fixing those pre-existing bugs seemed quite out-of-scope for this current PR, so I created a separate issue 989 [here](https://github.com/BioKIC/Symbiota/issues/989) to address those in the future.

**Comment by Atticus29** [comment]
> The underscores are showing up in the user interface now at the beginning of some of the search fields image

This is super weird. I can't reproduce this issue on my Chrome browser nor on Firefox at dev001.symbiota.org. Are you still seeing it?

**Comment by themerekat** [comment]
> > The underscores are showing up in the user interface now at the beginning of some of the search fields image
> 
> This is super weird. I can't reproduce this issue on my Chrome browser nor on Firefox at dev001.symbiota.org. Are you still seeing it?

Hm, ok now it looks like this:
image

**Comment by Atticus29** [comment]
> > When you conduct a search, then return to your Search Criteria using the breadcrumbs: image
> > All of your search criteria go away. In harvestparams, they stuck around so that you could refine your search instead of having to re-create it from scratch. I think this is a feature that people will want to keep.
> > That being said, I'm super excited to see that clicking the "back" button finally takes you to the search page without the scary "resubmit data" warning!
> 
> Ok. I fixed this in [0fd7569](https://github.com/BioKIC/Symbiota/commit/0fd7569d1b492106cbadbf0de5e73cc1dcfe495c). However, in the process of troubleshooting, I noticed some bugs that already existed in the harvestparams version as well. Fixing those pre-existing bugs seemed quite out-of-scope for this current PR, so I created a separate issue 989 [here](https://github.com/BioKIC/Symbiota/issues/989) to address those in the future.

This fix has been deployed on dev001.symbiota.org now.

**Comment by Atticus29** [comment]
> > > The underscores are showing up in the user interface now at the beginning of some of the search fields image
> > 
> > 
> > This is super weird. I can't reproduce this issue on my Chrome browser nor on Firefox at dev001.symbiota.org. Are you still seeing it?
> 
> Hm, ok now it looks like this: image

How do the other fields look? Like southern-latitude, etc.?

**Comment by themerekat** [comment]
> > > > The underscores are showing up in the user interface now at the beginning of some of the search fields image
> > > 
> > > 
> > > This is super weird. I can't reproduce this issue on my Chrome browser nor on Firefox at dev001.symbiota.org. Are you still seeing it?
> > 
> > 
> > Hm, ok now it looks like this: image
> 
> How do the other fields look? Like southern-latitude, etc.?

Similar: 
image

**Comment by Atticus29** [comment]
> > > > > The underscores are showing up in the user interface now at the beginning of some of the search fields image
> > > > 
> > > > 
> > > > This is super weird. I can't reproduce this issue on my Chrome browser nor on Firefox at dev001.symbiota.org. Are you still seeing it?
> > > 
> > > 
> > > Hm, ok now it looks like this: image
> > 
> > 
> > How do the other fields look? Like southern-latitude, etc.?
> 
> Similar: image

Ok. I'm working on this for 30 min. If I can't solve it in that timebox, I'm gonna create a separate ticket for it and kick it down the road.

**Comment by Atticus29** [comment]
> > > > > > The underscores are showing up in the user interface now at the beginning of some of the search fields image
> > > > > 
> > > > > 
> > > > > This is super weird. I can't reproduce this issue on my Chrome browser nor on Firefox at dev001.symbiota.org. Are you still seeing it?
> > > > 
> > > > 
> > > > Hm, ok now it looks like this: image
> > > 
> > > 
> > > How do the other fields look? Like southern-latitude, etc.?
> > 
> > 
> > Similar: image
> 
> Ok. I'm working on this for 30 min. If I can't solve it in that timebox, I'm gonna create a separate ticket for it and kick it down the road.

Ok @themerekat I changed up the styling completely. What do you think now?

**Comment by themerekat** [comment]
> > Ok. I'm working on this for 30 min. If I can't solve it in that timebox, I'm gonna create a separate ticket for it and kick it down the road.
> 
> Ok @themerekat I changed up the styling completely. What do you think now?

Hm, I don't like that it's not consistent with the rest of the fields. I prefer the previous version.

**Comment by Atticus29** [comment]
> > > Ok. I'm working on this for 30 min. If I can't solve it in that timebox, I'm gonna create a separate ticket for it and kick it down the road.
> > 
> > 
> > Ok @themerekat I changed up the styling completely. What do you think now?
> 
> Hm, I don't like that it's not consistent with the rest of the fields. I prefer the previous version.

Ok. I'll revert and just make an issue. Thanks!

**Comment by themerekat** [comment]
@Atticus29 , thanks, and sorry for the hassle!

**Comment by Atticus29** [comment]
> @Atticus29 , thanks, and sorry for the hassle!

No worries. Sorry it's such a weird issue. One day I just need to sit down for a few hours and fix it. But maybe bigger fish before 3.1 release. Ok, it's been reverted and re-deployed on dev001.

Issue filed as [991](https://github.com/BioKIC/Symbiota/issues/991).

**Comment by themerekat** [comment]
If I search by Scientific Name, I get the same results as when I search by Common Name. The search below should probably not produce any results:
image

HOWEVER, I just tested this in CCH2, and got the same results, so this isn't a bug (feature?) limited to the dev environment. Can probably be ignored.

**Review by themerekat — APPROVED** [review]

**Pull request merged**